collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java

```diff
@@ -0,0 +1,284 @@
+package org.odk.collect.android.formentry.backgroundlocation;
+
+import android.location.Location;
+
+import androidx.annotation.NonNull;
+
+import com.google.android.gms.location.LocationListener;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.location.client.LocationClient;
+import org.odk.collect.android.logic.AuditConfig;
+import org.odk.collect.android.logic.AuditEvent;
+
+/**
+ * Manages background location for the location audit logging and odk:setlocation action features.
+ * Provides precondition checking and user feedback for both features.
+ *
+ * For location audit logging, manages all the audit logging as well as fetching the location using
+ * Google Play Services.
+ *
+ * {@link org.odk.collect.android.logic.actions.setlocation.CollectSetLocationAction} fetches
+ * location for odk:setlocation actions.
+ *
+ * The implementation uses a state machine concept. Public method represent user or system actions
+ * that clients of this class capture. Based on those actions and various preconditions (Google Play
+ * Services available, location permissions granted, etc), the manager's state changes.
+ */
+public class BackgroundLocationManager implements LocationClient.LocationClientListener, LocationListener {
+    @NonNull
+    private BackgroundLocationState currentState;
+
+    @NonNull
+    private LocationClient locationClient;
+
+    @NonNull
+    private LocationListener locationListener;
+
+    @NonNull
+    private BackgroundLocationHelper helper;
+
+    public BackgroundLocationManager(LocationClient locationClient, BackgroundLocationHelper helper) {
+        currentState = BackgroundLocationState.NOT_APPLICABLE;
+        this.locationClient = locationClient;
+
+        this.helper = helper;
+
+        this.locationListener = this;
+    }
+
+    public void formFinishedLoading() {
+        if (currentState == BackgroundLocationState.NOT_APPLICABLE) {
```

Comment:
Maybe diving to deep into the state machine pattern but to me it would feel more consistent to have all these methods use a `switch`. In my head that makes it easy to skim through and see what states the method cares about and what it does in the default case.

Reply:
Agreed, will update. At a high-level, are you upset by the fact that the state transitions happen based not only on actions but also by various state checks triggered by the actions? For example, a more pure state machine design here might be to have the type of form checked outside this class and the event be something like `backgroundLocationFormFinishedLoading`. 

There might then also be a series of states between `APPLICABLE` (to be renamed) and `COLLECTING_LOCATION` instead of all the checks in `activityDisplayed` and `locationPermissionsGranted`.

I feel like what I ended up with strikes a good balance between grouping expected checks together and reaping the benefits of a state-machine-like structure but would like feedback on that.

Reply:
Hmmm. I think it's probably good enough for now as what it does nicely is pull knowledge of those transitions away from the Activity which doesn't care about them.

Reply:
👍 And I think there's a reasonable way to describe this model if we want to follow it elsewhere -- introduce events/state transitions for changes that only the activity can detect and push everything else into the model objects.